addons/website_sale/controllers/main.py

```diff
@@ -1225,3 +1225,56 @@ def products_autocomplete(self, term, options={}, **kwargs):
                 res_product['price'] = FieldMonetary.value_to_html(res_product['price'], monetary_options)
 
         return res
+
+    # --------------------------------------------------------------------------
+    # Products Recently Viewed
+    # --------------------------------------------------------------------------
+    @http.route('/shop/products/recently_viewed', type='json', auth='public', website=True)
+    def products_recently_viewed(self, **kwargs):
+        """
+        Returns list of recently viewed products according to current user and product options
+        """
+        excluded_ids = []
+        current_sale_order = request.session.get('sale_order_id')
+        if current_sale_order:
+            order_lines = request.env['sale.order.line'].sudo().search([('order_id', '=', current_sale_order)])
+            excluded_ids += order_lines.mapped('product_id.id')
+        if not request.env.user._is_public():
+            partner_id = request.env.user.partner_id
+            visitor_sudo = request.env['website.visitor'].sudo().with_context(active_test=False).search([('user_partner_id', '=', partner_id.id)])
+        else:
+            visitor_sudo = request.env['website.visitor']._get_visitor_from_request()
```

Comment:
This is intended, I want all of the visitor's previous visits, if he changes computer I want to display all the last product viewed products linked to him.

Reply:
Ah yeah indeed, you get all the products of all the found visitors for the user.
So I would call visitor_sudo -> 'visitors' + products -> 'visitors_products' so that we know that it can be multiple visitors, not just one.
Also, maybe explain a bit more in the doc string : 'Retrieve all the products see by the visitor or the user (from all user's related visitors' or something

Reply:
Good point about the need of doc str and better var name. Thx !